recent_topics: Standardize format of last message time.

amanagr opened the pull request:
We follow how other apps present older messages, e.g. Gmail,
Facebook Messenger, etc. display it.

Specifically, the logic we use is:

If the time is <24hr ago, show an absolute time, like "21:30" (or "9:30pm").
Otherwise, show what day it was, and not a time
  If the day was yesterday, say "Yesterday".
  Otherwise, if it was <7 days ago, say the day of week, like "Friday".
  Otherwise, if it was <1 year ago, say the month and day, like "Sep 6".
  Otherwise, say the year, month, and day, like "Sep 9, 2020".

Fixes #19775

gnprice reviewed (commented):
Thanks @amanagr! Comments below.

As I mentioned [in the chat thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/recent.20topics.20timestamps/near/1258873), I'd also love to see some test cases where:
> * Time zone is not UTC, and date of message / date of today / number of days ago are different in local time zone vs. UTC. Test that logic for which day of week to show, and whether to show day of week or date, correctly use the local time zone and don't get confused by mixing in UTC dates. (I don't see a sign that it would, but time-zone behavior should be tested, and it looks like the tests only mention UTC.)

amanagr commented on static/js/timerender.ts:
```diff
@@ -304,6 +305,26 @@ export function stringify_time(time: number | Date): string {
+        time_str = format(time, "MMM\u00A0dd,\u00A0yyyy");
+    } else {
+        time_str = format(time, "MMM\u00A0dd");
+    }
```
Oh! The last one is in else to handle if we get a message in future time too (unlikely but handled).

amanagr commented:
@gnprice thanks for the awesome review. I think I addressed all your comments in the way you wanted to. Left [one](https://github.com/zulip/zulip/pull/19830#discussion_r719472086) reply.

amanagr commented:
CI failure due to error in installing dependencies. 

```
Err:9 https://packages.groonga.org/ubuntu bionic Release
  Certificate verification failed: The certificate is NOT trusted. The certificate chain uses expired certificate.  Could not handshake: Error in the certificate verification. [IP: 153.126.209.250 443]
```

andersk commented on static/js/timerender.ts:
```diff
@@ -304,6 +305,28 @@ export function stringify_time(time: number | Date): string {
+    const days_old = differenceInCalendarDays(today, time);
+
+    const previous_year_leap = (today.getFullYear() - 1) % 4 === 0;
+    // If previous year is a leap year, same day will be 365 days ago and
+    // we want to display year in that case.
+    const a_year_has_passed = previous_year_leap ? days_old > 365 : days_old > 364;
```
This logic is incorrect in multiple ways.

* January 15, 2019 + 365 days = January 15, 2020
* January 15, 2020 + 366 days = January 15, 2021
* March 15, 2019 + 366 days = March 15, 2020
* March 15, 2020 + 365 days = March 15, 2021

timabbott commented:
@amanagr it looks like you missed the last batch of feedback from @gnprice?

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -15,8 +15,33 @@ user_settings.twenty_four_hour_time = true;
+function verify_timestamp(timestamp, time_ISO, DOW, UTC_OFFSET) {
+    const time = new Date(timestamp);
+    assert.equal(time.toISOString(), time_ISO);
+    assert.equal(new Intl.DateTimeFormat("en-US", {weekday: "long"}).format(time), DOW);
```
Thanks!

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -15,8 +15,33 @@ user_settings.twenty_four_hour_time = true;
+const timestamp_2019 = 1555091573000;
+
+// Wednesday, January 26, 2021 1:53 AM (UTC+0)
+const timestamp_2021 = 1611712388000;
```
I did the first point. I feel like point 2&3 would make it much harder to read code. Reading strings of time is harder to me than just constant variables.

gnprice replied:
Continued below: https://github.com/zulip/zulip/pull/19830#discussion_r800987863

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -15,8 +15,33 @@ user_settings.twenty_four_hour_time = true;
+    const time = new Date(timestamp);
+    assert.equal(time.toISOString(), time_ISO);
+    assert.equal(new Intl.DateTimeFormat("en-US", {weekday: "long"}).format(time), DOW);
+    assert.equal(timerender.get_tz_with_UTC_offset(time), UTC_OFFSET);
```
Okay done! Makes sense.

gnprice replied:
Thanks! Part of this is still open -- continued below: https://github.com/zulip/zulip/pull/19830#discussion_r800978283

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -89,8 +123,118 @@ run_test("render_now_returns_month_and_day", () => {
+
+run_test("format_time_modern_different_timezones", () => {
+    // This test checks if today and message time are in different timezones and have different
+    // dates / days ago in different timezones, then output time is correctly formatted.
```
Testing that timestamp is converted correctly in respective timezones is actually not the goal of this PR :).

I modified the tests to make it clear that we are testing that conversion of timestamp to formatted text works well in different timezones as well as in just `UTC+0`.

gnprice replied:
Continued below: https://github.com/zulip/zulip/pull/19830#discussion_r800992551

amanagr commented on static/js/timerender.ts:
```diff
@@ -304,6 +305,26 @@ export function stringify_time(time: number | Date): string {
+
+    if (days_old < 0) {
+        /* For timestamps in the future, we always show the year*/
+        return format(time, "MMM\u00A0dd,\u00A0yyyy");
```
Good catch! I fixed it to directly compare the timestamps, which should work even for a difference in seconds.

gnprice replied:
Continued below: https://github.com/zulip/zulip/pull/19830#discussion_r800998538

amanagr commented:
Thanks again for the detailed review. Sorry, this got delayed due to my personal problems.

gnprice reviewed (commented):
Thanks @amanagr for the revision! I didn't see this until just on Friday, sorry.

Comments below. I think at this point the behavior is all good, but there are still some things open in the tests, including some cases I'd like to have tested.

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -15,10 +15,33 @@ user_settings.twenty_four_hour_time = true;
+    const time = new Date(timestamp);
+    assert.equal(time.toISOString(), time_ISO);
+    assert.equal(new Intl.DateTimeFormat("en-US", {weekday: "long"}).format(time), DOW);
+    // Check there is no UTC offset.
     assert.equal(timerender.get_tz_with_UTC_offset(time), "UTC");
```
Okay, makes sense.

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -15,10 +15,33 @@ user_settings.twenty_four_hour_time = true;
     assert.equal(timerender.get_tz_with_UTC_offset(time), "UTC");
+}
+
+const timestamp_2017 = 1495091573000;
+verify_timestamp(timestamp_2017, "2017-05-18T07:12:53.000Z", "Thursday");
```
Yeah, make sense. I was hesitant to this change because we also timestamp in a lot of places. Using date.getTime() for that now.

amanagr commented on static/js/timerender.ts:
```diff
@@ -304,6 +305,26 @@ export function stringify_time(time: number | Date): string {
+    const hours = differenceInHours(today, time);
+    const days_old = differenceInCalendarDays(today, time);
+
+    if (time > today.getTime()) {
```
Yeah, it works that way too!

amanagr commented on frontend_tests/node_tests/timerender.js:
```diff
@@ -89,8 +117,83 @@ run_test("render_now_returns_month_and_day", () => {
+        timerender.format_time_modern(more_than_24_hours_ago, today),
+        "translated: Yesterday",
+    );
+    assert.equal(timerender.format_time_modern(less_than_a_week_ago, today), "Wednesday");
```
Hmm, ok added those!

amanagr commented:
I addressed all your comments, Greg (Thanks for the review again!). I reduced the commits to just 2 since it was very time-consuming to maintain all the commits with changes.

gnprice reviewed (approved):
All now looks good! Thanks again for implementing this and for all the revisions. Looking forward to seeing it in.

timabbott commented:
Merged, thanks @amanagr!
